crates/runtime/src/instance/allocator/pooling/uffd.rs

```diff
@@ -0,0 +1,833 @@
+
+const WASM_PAGE_SIZE: usize = wasmtime_environ::WASM_PAGE_SIZE as usize;
+
+pub unsafe fn make_accessible(_addr: *mut u8, _len: usize) -> bool {
+    // A no-op when userfaultfd is used
+    true
+}
+
+pub unsafe fn reset_guard_page(addr: *mut u8, len: usize) -> bool {
+    // Guard pages are READ_WRITE with uffd until faulted
+    region::protect(addr, len, region::Protection::READ_WRITE).is_ok()
+}
+
+pub unsafe fn decommit(addr: *mut u8, len: usize) {
+    // Use MADV_DONTNEED to mark the pages as missing
+    // This will cause a missing page fault for next access on any page in the given range
+    assert_eq!(
+        libc::madvise(addr as _, len, libc::MADV_DONTNEED),
+        0,
+        "madvise failed to mark pages as missing: {}",
+        std::io::Error::last_os_error()
+    );
+}
+
+pub fn create_memory_map(_accessible_size: usize, mapping_size: usize) -> Result<Mmap, String> {
+    // Allocate a single read-write region at once
+    // As writable pages need to count towards commit charge, use MAP_NORESERVE to override.
+    // This implies that the kernel is configured to allow overcommit or else
+    // this allocation will almost certainly fail without a plethora of physical memory to back the allocation.
+    // The consequence of not reserving is that our process may segfault on any write to a memory
+    // page that cannot be backed (i.e. out of memory conditions).
+
+    if mapping_size == 0 {
+        return Ok(Mmap::new());
+    }
+
+    unsafe {
+        let ptr = libc::mmap(
+            ptr::null_mut(),
+            mapping_size,
+            libc::PROT_READ | libc::PROT_WRITE,
+            libc::MAP_PRIVATE | libc::MAP_ANON | libc::MAP_NORESERVE,
+            -1,
+            0,
+        );
+
+        if ptr as isize == -1_isize {
+            return Err(format!(
+                "failed to allocate pool memory: {}",
+                std::io::Error::last_os_error()
+            ));
+        }
+
+        Ok(Mmap::from_raw(ptr as usize, mapping_size))
+    }
+}
+
+/// Represents a location of a page fault within monitored regions of memory.
+enum AddressLocation<'a> {
+    /// The address location is in a WebAssembly table page.
+    /// The fault handler will zero the page as tables are initialized at instantiation-time.
+    TablePage {
+        /// The address of the page being accessed.
+        page_addr: *mut u8,
+        /// The length of the page being accessed.
+        len: usize,
+    },
+    /// The address location is in a WebAssembly linear memory page.
+    /// The fault handler will copy the pages from initialization data if necessary.
+    MemoryPage {
+        /// The address of the page being accessed.
+        page_addr: *mut u8,
+        /// The length of the page being accessed.
+        len: usize,
+        /// The instance related to the memory page that was accessed.
+        instance: &'a Instance,
+        /// The index of the memory that was accessed.
+        memory_index: usize,
```

Comment:
Should this perhaps be wired up as `DefinedMemoryIndex`?

Reply:
Sure, that conveys the intention better.  I'll fix.